pkg/ingester/ingester_v2_backfill.go

```diff
@@ -0,0 +1,498 @@
+			app = db.Appender()
+			a.firstAppender = &backfillRangeAppender{app: app, db: db, start: start, end: end}
+		} else {
+			if a.backfillTSDB.dbs[1] != nil {
+				// This should not happen.
+				return nil, nil, errors.New("cannot create another backfill TSDB, older TSDB already exists")
+			}
+			a.backfillTSDB.dbs[1] = newDB
+			app = db.Appender()
+			a.secondAppender = &backfillRangeAppender{app: app, db: db, start: start, end: end}
+		}
+
+		a.ingester.metrics.numBackfillTSDBsPerUser.WithLabelValues(a.backfillTSDB.userID).Inc()
+		if a.backfillTSDB.dbs[0] == nil || a.backfillTSDB.dbs[1] == nil {
+			// This user did not have a backfill TSDB before.
+			a.ingester.metrics.numUsersWithBackfillTSDBs.Inc()
+		}
+	}
+	if app == nil {
+		return nil, nil, storage.ErrOutOfBounds
+	}
+
+	return app, db, nil
+}
+
+func (a *backfillAppender) commit() error {
+	var merr tsdb_errors.MultiError
+	if a.firstAppender != nil {
+		merr.Add(a.firstAppender.app.Commit())
+	}
+	if a.secondAppender != nil {
+		merr.Add(a.secondAppender.app.Commit())
+	}
+	return merr.Err()
+}
+
+func (a *backfillAppender) rollback() error {
+	var merr tsdb_errors.MultiError
+	if a.firstAppender != nil {
+		merr.Add(a.firstAppender.app.Rollback())
+	}
+	if a.secondAppender != nil {
+		merr.Add(a.secondAppender.app.Rollback())
+	}
+	return merr.Err()
+}
+
+func (u *userTSDB) backfillSelect(ctx context.Context, from, through int64, matchers []*labels.Matcher) ([]storage.SeriesSet, error) {
+	var queriers []storage.Querier
+	defer func() {
+		for _, q := range queriers {
+			q.Close()
+		}
+	}()
+
+	u.backfillTSDB.mtx.RLock()
+	for _, db := range u.backfillTSDB.dbs {
+		if db != nil && overlapsOpenInterval(db.start, db.end, from, through) {
+			mint := db.db.Head().MinTime()
+			maxt := db.db.Head().MaxTime()
+			if overlapsOpenInterval(mint, maxt, from, through) {
+				q, err := db.db.Querier(ctx, from, through)
+				if err != nil {
+					u.backfillTSDB.mtx.RUnlock()
+					return nil, err
+				}
+
+				queriers = append(queriers, q)
+			}
+		}
+	}
+	u.backfillTSDB.mtx.RUnlock()
+
+	if len(queriers) == 0 {
+		return nil, nil
+	}
+
+	result := make([]storage.SeriesSet, len(queriers))
+	for i, q := range queriers {
+		ss := q.Select(false, nil, matchers...)
+		if ss.Err() != nil {
+			return nil, ss.Err()
+		}
+		result[i] = ss
+	}
+
+	return result, nil
+}
+
+// compactAndShipAndDelete compacts, ships, closes, and deletes any TSDB that has
+// gone beyond the backfill age. Error in deletion of the TSDB directory is not returned.
+// force=true does the above operation irrespective of backfill age.
+func (b *backfillTSDB) compactAndShipAndDelete(force bool) (err error) {
+	b.compactMtx.Lock()
+	defer b.compactMtx.Unlock()
+
+	var merr tsdb_errors.MultiError
+	secondCompacted := false
+	for i := range []int{1, 0} {
+		b.mtx.RLock()
+		db := b.dbs[i]
+		b.mtx.RUnlock()
+
+		if db == nil || (!force && db.end > mtime.Now().Add(-b.backfillAge-time.Hour).Unix()*1000) {
```

Comment:
`db.end > mtime.Now().Add(-b.backfillAge-time.Hour).Unix()*1000` -- similar check repeats multiple times around the code... can we move it into a named function to explain its meaning?

Reply:
Definitely! Not sure why it was not already :)